pkg/scheduler/internal/queue/equivalence/eqivalence.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file contains structures that implement equivalence class types.
+// Pods under same controllerRef will be put in the same Class.podSet
+// and will be considered as an unit in the process of scheduling.
+// Equivalence class improves scheduler velocity and responsiveness by
+// continuous scheduling when one is determined schedulable and avoiding checking
+// the schedulability of all of these pods when one is determined unschedulable.
+// This is particularly useful in batch processing.
+
+package equivalence
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/scheduler/util"
+)
+
+// ClassMap is the type of a map which has the key by a pod's UID of controllerRef and
+// the value which is a pointer to the Class.
+type ClassMap map[types.UID]*Class
+
+var (
+	once     sync.Once
+	classMap ClassMap
+)
+
+// NewClassMap creates an empty classMap which is a map key by a pod's UID of
+// controllerRef and the value is a pointer to the Class.
+func NewClassMap() ClassMap {
+	// Make sure ClassMap is a singleton
+	once.Do(func() {
+		classMap = make(ClassMap)
+	})
+	return classMap
+}
+
+// Class holds pods under same controllerRef and is the unit in scheduling process.
+// of scheduling.
+type Class struct {
+	// Pods's UID of controllerRef
+	Hash types.UID
+	// Pods which under same controllerRef waiting to be scheduled.
+	PodSet *sync.Map
```

Comment:
Because PodSet is also operated in scheduling cycle. In order to prevent data race, so we use `sync.Map`.
[https://github.com/kubernetes/kubernetes/blob/50322a92ad38deddc6d293c9291dc316fcdbe010/pkg/scheduler/scheduler.go#L592-L608](url)
```go
equivalenceClass.PodSet.Range(func(k, v interface{}) bool {
			// Because Range() accesses read-only snapshots, it's possible that the previous
			// pod's scheduling time is too long to cause data to expire, so choose
			// to get the current data directly through Load().
			pod, ok := equivalenceClass.PodSet.Load(k)
			if !ok {
				return true
			}

			// If the pod is scheduled to fail, doSchedule returns false and range stops
			// the iteration.
			if !sched.doSchedule(pod.(*v1.Pod)) {
				return false
			}
			equivalenceClass.PodSet.Delete(pod.(*v1.Pod).UID)
			return true
		})
```